streams/test-utils/src/main/java/org/apache/kafka/streams/internal/KeyValueStoreFacade.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.internal;
+
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+import java.util.List;
+
+public class KeyValueStoreFacade<K, V> extends StoreFacade implements KeyValueStore<K, V> {
+    private final KeyValueStore<K, V> inner;
+    private final InternalProcessorContext context;
+
+    public KeyValueStoreFacade(final KeyValueStore<K, V> inner,
+                               final InternalProcessorContext context) {
+        super(inner);
+        this.inner = inner;
+        this.context = context;
+    }
+
+    @Override
+    public void put(final K key,
+                    final V value) {
+        setContext();
+        inner.put(key, value);
+        context.setRecordContext(null);
+    }
+
+    @Override
+    public V putIfAbsent(final K key,
+                         final V value) {
+        setContext();
+        final V oldValue = inner.putIfAbsent(key, value);
+        context.setRecordContext(null);
+        return oldValue;
+    }
+
+    @Override
+    public void putAll(final List<KeyValue<K, V>> entries) {
+        setContext();
+        inner.putAll(entries);
+        context.setRecordContext(null);
+    }
+
+    @Override
+    public V delete(final K key) {
+        setContext();
+        final V value = inner.delete(key);
+        context.setRecordContext(null);
+        return value;
+    }
+
+    private void setContext() {
```

Comment:
Could we move this to `StoreFacade`?

Reply:
After a second thought, I'm wondering if we can simply this a bit with the following (note this is a meta comment): 

After calling `task.initializeStateStores()` in `TopologyTestDriver` and get the context, call `setContext()` immediately to give it this dummy record context. Note that by doing this we will have sufficient information for callers before processing the first record. And once the first record is being processed the record context should have enough context already.

Admittedly it will not resolve the case that after all the records being processed, if `store.put()` gets called again it will be the same issue since internally we set `null` to record context. But I think we would not allow that operations any ways, so it should be fine?

Reply:
Our idea has the advantage that it avoids the casting issue (cf. https://github.com/apache/kafka/pull/5096#discussion_r191643617) and also avoids all the code for the store-facades...

However, using the store-facades we are able to protect the store from invalid calls (like init() and close()) and can also throw better error messages. Also, calling put() after processing starts might put "undesired" data into the changelog topic, because some metadata will be inherited from the last processed record...

An open question: should we allow to call `put()` after processing started? IMHO, not, because the idea is to pre-populate the store during test setup. We can disallow this by setting the recordContext to `null` before `pipeInput` returns. However, the error message will be rather un-intuitive:
```
java.lang.IllegalStateException: This should not happen as timestamp() should only be called while a record is processed
```
Also, if change-logging is disabled, no exception would be thrown.

If we want to be more strict about when users are allowed to put data into a store from external, I think we need the facades.

I am not sure atm what the better/cleaner solution is, and what might provide the better user experience -- also, how important it is to guard store access. To me, it seems good to limit store access, because otherwise people might shoot themselves in the foot because they might manipulate a store mid-test (what can never happen in reality) and thus write an incorrect test.

Let me know what you think. (Also \cc @wushujames )

Reply:
I think that it makes sense to disallow people from manipulating the state store after the topology has started. As you point out, once the topology is started, all updates to the state store should happen via the topology itself. The only time we should manipulate the state store directly would be to pre-populate it.

So, disallowing calls to `put()` would prevent people from shooting themselves in the foot. Similarly, disallowing calls to `init()` and `close()` would also prevent people from shooting themselves in the foot.

So I think having the facade's makes sense.

Reply:
Btw, to give some background about why I was manipulating state stores in the first place...

I was testing how my topology behaves when it is updating multiple state stores and then crashes. With exactly_once turned on, it should behave fine, since the changes will be atomic - either the transaction will succeed, or it will be rolled back.

With exactly_once turned off, I believe it's possible that updates to one state store will succeed while another will fail. (Can any of you confirm this?) If I have 2 state stores, and I write to both and crash, I won't know which one succeeded. I was manipulating the state stores so that I can run tests where one state store gets updated and the other doesn't, and then can run my topology to test that both converge to the same state. That would allow me to be (more?) confident that the application will work correctly without exactly_once transactions.

It would be an interesting RFE if the `TopologyTestDriver` could do that testing for me. With 2 state stores, there are 4 possible scenarios I need to test. With 3 state stores, there are 8 possible scenarios I need to test. It would be great if somehow the TopologyTestDriver could notice the contents of the state store prior to processing, notice the contents of the state store after processing, and then re-run the processing of the record with all possible combinations of state-store updates, and make sure that all converge to the same state.

Reply:
Speaking of limiting access to state store APIs, I can apparently call `init()` and `close()` when using a state store within a `Transformer`. Should that also be disallowed? (If so, we'd do that in a separate PR)

Reply:
@wushujames You cannot use the test driver for EOS... For this, you would need to write an integration test using an embedded broker -- embedded broker is no public API atm, but I hope it will be added eventually. For now, you would need to add the test package into your dependency scope the get access to the embedded broker.

The issue wish `init()` and `close()` in `Transform` (and `TransformValues` and `Processor`) is valid. Please create a JIRA for this -- up to know, it's only documented in JavaDocs to not close the stores AFAIK. Would be good to get some guards for both.

Reply:
Right, understood about the `TopologyTestDriver` not being used to EOS. I was actually using it to test that things would still work properly when I turned *off* `EOS`.

I will open up a JIRA about `init()` and `close()`.

Reply:
About EOS on/off: it make actually not difference, because the test driver cannot be used to test error scenarios, but only the happy path. And as long as no error occurs, there won't be duplicates anyway.

Reply:
Filed 
[KAFKA-6970] Kafka streams lets the user call init() and close() on a state store, when inside Processors
https://issues.apache.org/jira/browse/KAFKA-6970

Reply:
I wanted to disassembly our discussions into three topics, which to me are semi-separated:

1. Allowing the state stores to be pre-populated before processing any records, this is what KAFKA-6967 is originally filed for.

2. Disallowing store APIs that are supposed to be only used by library, e.g. `init()` and `close()`. Note that this is not only in TopologyTestDriver, but more generally in KafkaStreams actual runtime, e.g. as we mentioned in `Transformer` and `Processor`. I think a general solution would be in the wrapper store, e.g. `MeteredXXStore`, double check that the store is `initialized`, `closed` etc so that double calling it will result in exception.

3. Disallowing store.puts in between of processing a record. Again this is more general than just the TopologyTestDriver itself: imagine if users could get the store directly with reflections (with the current API they can only get ReadOnlyXXStore, so that would not be possible), then they can still call puts without problems. Likewise, I think it worth a general solution inside StreamTask, e.g. to set the record context to null, similar to processor context so that `put` call will throw.

As for execution, I'd suggest we do not use store facade but just set record context in TopologyTestDriver constructor, in order to just fix 1) above, and defer 2) and 3) to future works as they are more general fixes that may have larger impact. As we are approaching the release deadline I'd prefer not coupling those larger changes in this single PR.

Reply:
@guozhangwang: I'm fine with fixing 1), and then deferring 2) and 3).